src/Swift.Runtime/src/Swift/Data.cs

```diff
@@ -1,10 +1,119 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Swift;
+using Swift.Runtime;
+
+
 /// <summary>
-/// Represents Foundation.Data type.
+/// Represents Swift Foundation.DataProtocol in C#.
+/// </summary>
+public interface ISwiftDataProtocol { }
+
+/// <summary>
+/// Represents Swift Foundation.ContiguousBytes in C#.
 /// </summary>
+public interface ISwiftContiguousBytes { }
 
-// TODO: https://github.com/dotnet/runtimelab/issues/2992
-public struct Data
+/// <summary>
+/// Represents Foundation.Data type.
+/// </summary>
+public unsafe struct Data : ISwiftObject
 {
     private long _flags;
     private IntPtr _object;
+
+    private static nuint _payloadSize = SwiftObjectHelper<Data>.GetTypeMetadata().Size;
+
+    private static Dictionary<Type, string> _protocolConformanceSymbols;
+
+    static Data()
+    {
+        _protocolConformanceSymbols = new Dictionary<Type, string> {
+            { typeof(ISwiftDataProtocol), "$s10Foundation4DataVAA0B8ProtocolAAMc" },
+            { typeof(ISwiftContiguousBytes), "$s10Foundation4DataVAA15ContiguousBytesAAMc" },
+        };
+    }
+
+    static TypeMetadata ISwiftObject.GetTypeMetadata()
+    {
+        return TypeMetadata.Cache.GetOrAdd(typeof(Data), _ => PInvoke_getMetadata());
+    }
+
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+    [DllImport(KnownLibraries.SwiftFoundation, EntryPoint = "$s10Foundation4DataVMa")]
+    public static unsafe extern TypeMetadata PInvoke_getMetadata();
+
+    static ISwiftObject ISwiftObject.NewFromPayload(SwiftHandle handle)
+    {
+        return *(Data*)handle;
+    }
+
+    IntPtr ISwiftObject.MarshalToSwift(IntPtr swiftDest)
+    {
+        var metadata = SwiftObjectHelper<Data>.GetTypeMetadata();
+        unsafe
+        {
+            fixed (void* _payloadPtr = &this)
+            {
+                metadata.ValueWitnessTable->InitializeWithCopy((void*)swiftDest, (void*)_payloadPtr, metadata);
```

Comment:
This is a broader question (not just Data): In `MarshalToSwift` we call InitializeWithCopy, when copying the content to a native memory buffer. But in the `NewFromPayload` (and in the constructor) we make a payload copy with just "memcpy". The question is why is it OK to do a simple memory copy when bringing the data into managed but it's not OK to do it when creating a native memory copy? It feels inconsistent.

Reply:
We've discussed this offline and it seems that the existing copy constructor is not valid.

Reply:
I'm sorry I was not able to be there for the discussion. Do we have some kind of writeup which describes the outcomes? We'll probably need some kind of rules to follow in other projections as well.

Reply:
Yes, sorry for the vague response. I was waiting for this PR to streamline the feedback: https://github.com/dotnet/runtimelab/pull/3008